pkg/controller/workload/mpijob/mpijob_controller.go

```diff
@@ -0,0 +1,579 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mpijob
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	kubeflow "github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v2beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/constants"
+	"sigs.k8s.io/kueue/pkg/controller/workload/jobframework"
+	utilpriority "sigs.k8s.io/kueue/pkg/util/priority"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+var (
+	ownerKey = ".metadata.mpijob_controller"
+)
+
+// MPIJobReconciler reconciles a Job object
+type MPIJobReconciler struct {
+	client                     client.Client
+	scheme                     *runtime.Scheme
+	record                     record.EventRecorder
+	manageJobsWithoutQueueName bool
+	waitForPodsReady           bool
+}
+
+type options struct {
+	manageJobsWithoutQueueName bool
+	waitForPodsReady           bool
+}
+
+// Option configures the reconciler.
+type Option func(*options)
+
+// WithManageJobsWithoutQueueName indicates if the controller should reconcile
+// jobs that don't set the queue name annotation.
+func WithManageJobsWithoutQueueName(f bool) Option {
+	return func(o *options) {
+		o.manageJobsWithoutQueueName = f
+	}
+}
+
+// WithWaitForPodsReady indicates if the controller should add the PodsReady
+// condition to the workload when the corresponding job has all pods ready
+// or succeeded.
+func WithWaitForPodsReady(f bool) Option {
+	return func(o *options) {
+		o.waitForPodsReady = f
+	}
+}
+
+var defaultOptions = options{}
+
+func NewReconciler(
+	scheme *runtime.Scheme,
+	client client.Client,
+	record record.EventRecorder,
+	opts ...Option) *MPIJobReconciler {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &MPIJobReconciler{
+		scheme:                     scheme,
+		client:                     client,
+		record:                     record,
+		manageJobsWithoutQueueName: options.manageJobsWithoutQueueName,
+		waitForPodsReady:           options.waitForPodsReady,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager. It indexes workloads
+// based on the owning jobs.
+func (r *MPIJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kubeflow.MPIJob{}).
+		Owns(&kueue.Workload{}).
+		Complete(r)
+}
```

Comment:
If there are no CRDs for MPIJob in the cluster, the kueue-controller always says logs in the following, although kueue works fine:

```shell
{"level":"error","ts":"2023-03-01T16:01:40.4642038Z","logger":"controller-runtime.source","caller":"source/source.go:143","msg":"if kind is a CRD, it should be installed before calling Start","kind":"MPIJob.kubeflow.org","error":"no matches for kind \"MPIJob\" in version \"kubeflow.org/v2beta1\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start.func1.1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.2/pkg/source/source.go:143\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:235\nk8s.io/apimachinery/pkg/util/wait.WaitForWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:662\nk8s.io/apimachinery/pkg/util/wait.poll\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:596\nk8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:547\nsigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start.func1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.2/pkg/source/source.go:136"}
```

Maybe we can skip to watch MPIJob using following library if there are no CRDs for MPIJob:

```go
// If err is not nil, there are no CRDs for MPIJob in the cluster.
_, err := mgr.GetRESTMapper().RESTMapping(schema.GroupKind{Group: kubeflow.SchemeGroupVersion.Group, Kind: "MPIJob"})
```

WDYT?

Reply:
I guess it's an overkill for now, as we plan to move this code to a separate binary

Reply:
> I guess it's an overkill for now, as we plan to move this code to a separate binary

Does that mean we run controllers for CustomJobs as sidecar containers?

Reply:
Either sidecars or separate deployments.

Reply:
sgtm

Reply:
I suggest this as a future improvement that may not be needed if we move the code out to a separate binary (probably next release though).